provider/reprovider.go

```diff
@@ -236,208 +302,223 @@ func initialReprovideDelay(duration time.Duration) Option {
+			case <-provideDelayTimer.C:
+				if !lateOnProvides {
+					log.Warn("New provides are piling up in the queue, consider increasing the number of provide workers or enabling Batch Providing.")
+					lateOnProvides = true
+				}
+				select {
+				case provideQueue <- c:
+				case <-ctx.Done():
+				}
+			case <-ctx.Done():
 			}
-
-			if r, ok := s.rsys.(Ready); ok {
-				ticker := time.NewTicker(time.Minute)
-				for !r.Ready() {
-					log.Debugf("reprovider system not ready")
+		}
+		// Start provide workers
+		for range s.provideWorkerCount {
+			go func(ctx context.Context) {
+				for {
 					select {
-					case <-ticker.C:
-					case <-s.ctx.Done():
+					case c := <-provideQueue:
+						provideFunc(ctx, c)
+					case <-ctx.Done():
 						return
 					}
 				}
-				ticker.Stop()
-			}
+			}(s.ctx)
+		}
+	}
 
-			log.Debugf("starting provide of %d keys", len(keys))
-			start := time.Now()
-			err := doProvideMany(s.ctx, s.rsys, keys)
-			if err != nil {
-				log.Debugf("providing failed %v", err)
+	for {
+		select {
+		case c := <-provCh:
+			if err := verifcid.ValidateCid(s.allowlist, c); err != nil {
+				log.Errorf("insecure hash in reprovider, %s (%s)", c, err)
 				continue
 			}
-			dur := time.Since(start)
-
-			totalProvideTime := time.Duration(s.totalProvides) * s.avgProvideDuration
-			recentAvgProvideDuration := dur / time.Duration(len(keys))
-
-			s.statLk.Lock()
-			s.avgProvideDuration = (totalProvideTime + dur) / (time.Duration(s.totalProvides) + time.Duration(len(keys)))
-			s.totalProvides += uint64(len(keys))
-
-			log.Debugf("finished providing of %d keys. It took %v with an average of %v per provide", len(keys), dur, recentAvgProvideDuration)
-
-			if performedReprovide {
-				s.lastReprovideBatchSize = uint64(len(keys))
-				s.lastReprovideDuration = dur
-				s.lastRun = time.Now()
-
-				s.statLk.Unlock()
-				// Don't hold the lock while writing to disk, consumers don't need to wait on IO to read thoses fields.
-
-				// persist last reprovide time to disk to avoid unnecessary reprovides on restart
-				if err := s.ds.Put(s.ctx, lastReprovideKey, storeTime(s.lastRun)); err != nil {
-					log.Errorf("could not store last reprovide time: %v", err)
-				}
-				if err := s.ds.Sync(s.ctx, lastReprovideKey); err != nil {
-					log.Errorf("could not perform sync of last reprovide time: %v", err)
-				}
-			} else {
-				s.statLk.Unlock()
-			}
-
-			s.throughputDurationSum += dur
-			s.throughputProvideCurrentCount += uint(len(keys))
-			if s.throughputCallback != nil && s.throughputProvideCurrentCount >= s.throughputMinimumProvides {
-				if more := s.throughputCallback(performedReprovide, complete, s.throughputProvideCurrentCount, s.throughputDurationSum); !more {
-					s.throughputCallback = nil
-				}
-				s.throughputProvideCurrentCount = 0
-				s.throughputDurationSum = 0
+			// Deduplicate recently provided cids
+			if found, _ := s.instantProvideDeduplicator.ContainsOrAdd(c, struct{}{}); found {
```

Comment:
Consider moving this into the provide queue (`s.q`). Doing that will prevent duplicate CIDs from having to sit in the queue. The  queue's internal buffer can be replaced with this LRU cache, and there is no need to make any other logic changes.

Reply:
done in https://github.com/ipfs/boxo/pull/909